go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -29,56 +31,237 @@ const (
+			}
+		}
+
+		// for arp request/response from vxlan, learn and add flow to table 20, for pod-> vxlan traffic
+		// special cookie needed here for tunnel
+		// tunnel cookie flows only contain a single flow ever, but it is updated by multiple pod adds
+		// so need proper locking around tunMap
+		// after learning actions, we need to resubmit the flow to the gw arp response table (2) so that we can respond
+		// back if this was an arp request
+		tunCookie := podIPToCookie(net.ParseIP(namespaceVTEP))
+		tunFlow := fmt.Sprintf("table=0,cookie=0x%s,priority=120,in_port=%s,arp,arp_spa=%s,tun_src=%s,"+
+			"actions=%s,resubmit(,2)",
+			tunCookie, extVXLANName, namespaceExternalGw, namespaceVTEP, learnActions)
+		n.updateFlowCacheEntry(tunCookie, []string{tunFlow}, false)
+		n.tunMapMutex.Unlock()
+
+		// add flow to table 0 to match on incoming traffic from pods, send to table 20
+		// bypass regular Hybrid overlay for gateway mode
+		flows = append(flows,
+			fmt.Sprintf("table=0, cookie=0x%s, priority=10000,in_port=ext,ip,nw_src=%s,"+
+				"actions=goto_table:20",
+				cookie, podIP.IP))
+
+		// we need to send an ARP request to get the GW to send us a response
+		// and learn the mac, we will trigger an arp request to the gateway in table 1
+		flows = append(flows,
+			fmt.Sprintf(""+
+				"table=1,cookie=0x%s,priority=10,arp,arp_tpa=%s,"+
+				"actions="+
+				"mod_dl_dst:ff:ff:ff:ff:ff:ff,"+
+				"mod_dl_src:%s,"+
+				"load:0x1->NXM_OF_ARP_OP[],"+
+				"set_field:%s->arp_sha,"+
+				"set_field:%s->arp_spa,"+
+				"set_field:%s->arp_tpa,"+
+				"set_field:00:00:00:00:00:00->arp_tha,"+
+				"load:%d->NXM_NX_TUN_ID[0..31],"+
+				"set_field:%s->tun_dst,"+
+				"output:%s",
+				cookie, podIP.IP, n.drMAC.String(), n.drMAC.String(), n.drIP, namespaceExternalGw, hotypes.HybridOverlayVNI,
+				namespaceVTEP, extVXLANName))
+		n.updateFlowCacheEntry(cookie, flows, ignoreLearn)
+	}
+	n.requestFlowSync()
+	klog.Infof("Pod %s wired for Hybrid Overlay", pod.Name)
 	return nil
 }
 
 func (n *NodeController) deletePod(pod *kapi.Pod) error {
 	if pod.Spec.NodeName == n.nodeName {
-		if err := deleteFlowsByCookie(10, podToCookie(pod)); err != nil {
-			return fmt.Errorf("failed to delete flows for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		podIPs, _, err := getPodDetails(pod, n.nodeName)
+		if err != nil {
+			return fmt.Errorf("error getting pod details: %v", err)
+		}
+		tunIPs := make(map[string]struct{})
+		n.tunMapMutex.Lock()
+		for _, podIP := range podIPs {
+			// need to check if any pods in the tunMap still correspond to a tunnel
+			// store the tunIP so we can delete cookie later
+			tunIPs[n.tunMap[podIP.IP.String()]] = struct{}{}
+			delete(n.tunMap, podIP.IP.String())
+		}
+		for tunIP := range tunIPs {
+			if len(tunIP) > 0 {
+				// check if any pods still belong to this tunnel so we can clean up the flow if not
+				tunStillActive := false
+				for _, tun := range n.tunMap {
```

Comment:
So I think 2 maps is not a good idea, and the reason this map uses this ordering is to prevent using slices. I figured the cost of copying the array over and over to build the slice was heavier than the alternative (as it is likely there are going to be a lot of pods pertaining to 1 tunnel). But maybe like you are saying the iteration cost is heavier than that. I'm not sure, but is the difference negligible either way and we should just leave it as is (since we know it works)?

Reply:
will leave it for now and we can optimize it later